### DISCUSSION TITLE
format in tab.text() not working as expected

### DISCUSSION THREAD

**Discussion by mauf3r** [Avaiga/taipy#1534] [Q&A]
Hi there,

I am trying to show some KPIs, inspired by what streamlit's st.metric() does.
And `tgb.text()` gives me some headache...

The basic Idea works with:
```pyhton
self.x = 12345,67890
with tgb.part('card'):
    tgb.text(f"**Headline**", mode='md')
    tgb.text("{x}")
```

**1st issue**: format does not work as expected
```pyhton
self.x = 12345,67890
with tgb.part('card'):
    tgb.text(f"**Headline**", mode='md')
    tgb.text("{x}")
    tgb.text("{x}", format="%.0f")
    tgb.text("{x}", format="%,.0f")
```
 **Result:**
12345.6789
12346
12.345,679

**2nd issue**: handling of expressions is not working as expected
```pyhton
self.x = {'a': {'b': {'': 12345.67890}}}
with tgb.part('card'):
    tgb.text(f"**Headline**", mode='md')
    tgb.text("{x.get('a',{}).get('b',{}).get('',0)")
```
taipy_app-1  |     x.get('a',{
taipy_app-1  |               ^
taipy_app-1  | SyntaxError: '{' was never closed

Fun-Fact 1:
```pyhton
 tgb.text("{x.get('a').get('b',{}).get('',0)")
 tgb.text("{x.get('a',{}).get('b').get('',0)")
```
works!

Fun-Fact 2:
using `x.get('a')` the format arg is completely ignored, with `x['a']` it works as described above

**Comment by FlorianJacta** [comment]
In fact, we have a [metric visual element](https://docs.taipy.io/en/develop/manuals/gui/viselements/metric/) that will be released in the next release. 

You can use our develop branch by `pip install git+https://github.com/Avaiga/taipy.git`. That will take more time than normal.

You have an issue in this sentence; you have not closed the `{` as the error is saying:

`tgb.text("{x.get('a',{}).get('b',{}).get('',0)")` -> `tgb.text("{x.get('a',{}).get('b',{}).get('',0)"})`

I suggest not having `{}` inside a bound expression/variable. You should create a function that taipy is calling. For example:

```python
def format_number(x):
    # do whatever you want here. When *state.x* changes, this function will be called.
    return x.get('a',{}).get('b',{}).get('',0)

self.x = {'a': {'b': {'': 12345.67890}}}
with tgb.part('card'):
    tgb.text(f"**Headline**", mode='md')
    tgb.text("{format_number(x)}")
```


I am not sure about the error with `"%,.0f"`.

**Reply by mauf3r** [reply]
Unfortunately, that seems to be one of the issues you mentioned.
So just for the records - unless I made a mistake elsewhere:

```python
    def convert(self, x):
        x_new =  x.get('a',{}).get('b',{}).get('',0)
        return f"{x_new:.,0f}"

    def home_page(self):
        with tgb.part("container"):
                with tgb.part('card'):
                    tgb.text("{convert(x)}", mode='md')
```

Does not give any output.

**Reply by FlorianJacta** [reply]
Try to define the function `convert(x)` globally without the *self* in the parameters.
